### PR TITLE
Validate build_daemon issues

### PR DESCRIPTION
**Pull request by grouma**
- Validate that the version of `build_daemon` used by webdev is compatible with the transitive version used by `build_runner`
- If invalid, check to see if an updated version of webdev would resolve the issue
- If not, tell the user to add a constraint on `build_daemon`

Somewhat towards: https://github.com/dart-lang/webdev/issues/114

### CODE REVIEW

File: webdev/lib/src/pubspec.dart [hunk 1]
```diff
@@ -125,7 +170,33 @@ Future<void> checkPubspecLock(PubspecLock pubspecLock,
...
+  final VersionConstraint buildDaemonConstraint;
+  final bool isNewer;
+  _PackageInfo(this.version, this.buildDaemonConstraint, this.isNewer);
+}
+
+/// Returns the package info for the latest webdev release.
+Future<_PackageInfo> _latestPackageInfo() async {
+  var response = await get('https://pub.dartlang.org/api/packages/webdev');
+  var responseObj = json.decode(response.body);
+  var pubVersionString = responseObj['latest']['pubspec']['version'] as String;
```
Comment (kevmoo) [inline webdev/lib/src/pubspec.dart]: I do the heavy lifting for you! https://pub.dartlang.org/documentation/pubspec_parse/latest/pubspec_parse/Pubspec/Pubspec.fromJson.html

File: webdev/lib/src/pubspec.dart [hunk 2]
```diff
@@ -125,7 +170,33 @@ Future<void> checkPubspecLock(PubspecLock pubspecLock,
...
+class _PackageInfo {
+  final Version version;
+  final VersionConstraint buildDaemonConstraint;
+  final bool isNewer;
+  _PackageInfo(this.version, this.buildDaemonConstraint, this.isNewer);
+}
+
+/// Returns the package info for the latest webdev release.
+Future<_PackageInfo> _latestPackageInfo() async {
+  var response = await get('https://pub.dartlang.org/api/packages/webdev');
```
Comment (kevmoo) [inline webdev/lib/src/pubspec.dart]: See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent on request. Include `webdev` and the version here...

File: webdev/lib/src/pubspec.dart [hunk 3]
```diff
@@ -113,6 +122,42 @@ class PubspecLock {
...
+
+  var issues = <PackageExceptionDetails>[];
+
+  var buildDaemonIssues = pubspecLock.checkPackage(
+    'build_daemon',
+    VersionConstraint.parse(buildDaemonConstraint),
+  );
+
+  // Only warn of build_daemon issues if they have a dependency on the package.
+  if (buildDaemonIssues.any((issue) => !issue._missingDependency)) {
```
Comment (jakemac53) [inline webdev/lib/src/pubspec.dart]: Wouldn't this version of webdev also require the build_daemon dependency though? Or will that be messaged by some other warning (the build_runner version check?)

File: webdev/lib/src/pubspec.dart [hunk 4]
```diff
@@ -125,7 +170,33 @@ Future<void> checkPubspecLock(PubspecLock pubspecLock,
...
+  _PackageInfo(this.version, this.buildDaemonConstraint, this.isNewer);
+}
+
+/// Returns the package info for the latest webdev release.
+Future<_PackageInfo> _latestPackageInfo() async {
+  var response = await get('https://pub.dartlang.org/api/packages/webdev');
+  var responseObj = json.decode(response.body);
+  var pubVersionString = responseObj['latest']['pubspec']['version'] as String;
+  var buildDaemonConfig = responseObj['latest']['pubspec']['dev_dependencies']
+      ['build_demon'] as String;
```
Comment (jakemac53) [inline webdev/lib/src/pubspec.dart]: daemon

File: webdev/lib/src/pubspec.dart (hunk 3 above)
Comment (grouma) [reply webdev/lib/src/pubspec.dart]: Ahh, I need to update the above check to set `requireDirect: false`

Comment (grouma) [reply webdev/lib/src/pubspec.dart]: On second thought it isn't required. I added the `requireDirect` anyway to make it explicit.

File: webdev/lib/src/pubspec.dart [hunk 5]
```diff
@@ -186,17 +188,19 @@ class _PackageInfo {
...
   var responseObj = json.decode(response.body);
-  var pubVersionString = responseObj['latest']['pubspec']['version'] as String;
-  var buildDaemonConfig = responseObj['latest']['pubspec']['dev_dependencies']
-      ['build_demon'] as String;
-  var buildDaemonConstraint = buildDaemonConfig != null
-      ? VersionConstraint.parse(buildDaemonConfig)
-      // This should never be satisfied.
-      : VersionConstraint.parse('0.0.0');
-  var pubVersion = Version.parse(pubVersionString);
+  var pubspec = Pubspec.fromJson(
```
Comment (kevmoo) [inline webdev/lib/src/pubspec.dart]: Ha! Didn't save you many lines!

Comment (grouma) [reply webdev/lib/src/pubspec.dart]: Yeah. I'm tempted to not depend on the package but don't feel too strongly.

Comment (jakemac53) [reply webdev/lib/src/pubspec.dart]: Might give better errors?

Comment (kevmoo) [reply webdev/lib/src/pubspec.dart]: Yeah. Better to avoid hand writing another parser.

**Pull request merged**